File: _sources/FireflyCore/Core/CommandLine.vb
'==========================================================================
'
'  File:        CommandLine.vb
'  Location:    Firefly.Core <Visual Basic .Net>
'  Description: 控制台
'  Version:     2010.01.31.
'  Copyright(C) F.R.C.
'
'==========================================================================

Option Strict On
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text.RegularExpressions

Public NotInheritable Class CommandLine
    Private Sub New()
    End Sub

    Public Class CommandLineOption
        Public Name As String
        Public Arguments As String()
    End Class

    Public Class CommandLineArguments
        Public Arguments As String()
        Public Options As CommandLineOption()
    End Class

    Private Shared Function DescapeQuota(ByVal s As String) As String
        If s = "" Then Return ""
        If s.StartsWith("""") AndAlso s.EndsWith("""") Then
            Return s.Substring(1, s.Length - 2).Replace("""""", """")
        Else
            Return s
        End If
    End Function

    Public Shared Function GetCmdLine() As CommandLineArguments
        Dim argv As New List(Of String)
        Dim SuppressedFirst As Boolean = False
        Dim NextStart = 0
        Dim CmdLine = Environment.CommandLine
        For Each arg As Match In Regex.Matches(CmdLine, "(""[^""]*""|([^"" ])+)+", RegexOptions.ExplicitCapture)
            If arg.Index <> NextStart Then
                If Not CmdLine.Substring(NextStart, arg.Index - NextStart).All(Function(c) c = " ") Then Throw New InvalidOperationException
            End If
            NextStart = arg.Index + arg.Length
            If Not SuppressedFirst Then
                SuppressedFirst = True
                Continue For
            End If
            If arg.Success Then
                Dim m = arg.Value
                argv.Add(DescapeQuota(m))
            Else
                Throw New InvalidOperationException
            End If
        Next
        If CmdLine.Length <> NextStart Then
            If Not CmdLine.Substring(NextStart, CmdLine.Length - NextStart).All(Function(c) c = " ") Then Throw New InvalidOperationException
        End If

        Dim Arguments As New List(Of String)
        Dim Options As New List(Of CommandLineOption)

        For Each arg In argv
            If arg.StartsWith("/") Then
                Dim OptionLine = arg.Substring(1)
                Dim Name As String
                Dim ParameterLine As String
                Dim Index = OptionLine.IndexOf(":")
                If Index >= 0 Then
                    Name = DescapeQuota(OptionLine.Substring(0, Index))
                    ParameterLine = OptionLine.Substring(Index + 1)
                Else
                    Name = DescapeQuota(OptionLine)
                    ParameterLine = ""
                End If
                Options.Add(New CommandLineOption With {.Name = Name, .Arguments = (From t In ParameterLine.Split(","c) Select DescapeQuota(t.Trim(" "c))).ToArray})
            Else
                Arguments.Add(arg)
            End If
        Next

        Return New CommandLineArguments With {.Arguments = Arguments.ToArray, .Options = Options.ToArray}
    End Function
End Class

File: _sources/FireflyCore/Core/CommandLine.cs
// ==========================================================================
// 
// File:        CommandLine.vb
// Location:    Firefly.Core <Visual Basic .Net>
// Description: 控制台
// Version:     2010.01.31.
// Copyright(C) F.R.C.
// 
// ==========================================================================

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using Microsoft.VisualBasic.CompilerServices;

namespace Firefly
{

    public sealed class CommandLine
    {
        private CommandLine()
        {
        }

        public class CommandLineOption
        {
            public string Name;
            public string[] Arguments;
        }

        public class CommandLineArguments
        {
            public string[] Arguments;
            public CommandLineOption[] Options;
        }

        private static string DescapeQuota(string s)
        {
            if (string.IsNullOrEmpty(s))
                return "";
            if (s.StartsWith("\"") && s.EndsWith("\""))
            {
                return s.Substring(1, s.Length - 2).Replace("\"\"", "\"");
            }
            else
            {
                return s;
            }
        }

        public static CommandLineArguments GetCmdLine()
        {
            var argv = new List<string>();
            bool SuppressedFirst = false;
            int NextStart = 0;
            string CmdLine = Environment.CommandLine;
            foreach (Match arg in Regex.Matches(CmdLine, "(\"[^\"]*\"|([^\" ])+)+", RegexOptions.ExplicitCapture))
            {
                if (arg.Index != NextStart)
                {
                    if (!CmdLine.Substring(NextStart, arg.Index - NextStart).All(c => Conversions.ToString(c) == " "))
                        throw new InvalidOperationException();
                }
                NextStart = arg.Index + arg.Length;
                if (!SuppressedFirst)
                {
                    SuppressedFirst = true;
                    continue;
                }
                if (arg.Success)
                {
                    string m = arg.Value;
                    argv.Add(DescapeQuota(m));
                }
                else
                {
                    throw new InvalidOperationException();
                }
            }
            if (CmdLine.Length != NextStart)
            {
                if (!CmdLine.Substring(NextStart, CmdLine.Length - NextStart).All(c => Conversions.ToString(c) == " "))
                    throw new InvalidOperationException();
            }

            var Arguments = new List<string>();
            var Options = new List<CommandLineOption>();

            foreach (var arg in argv)
            {
                if (arg.StartsWith("/"))
                {
                    string OptionLine = arg.Substring(1);
                    string Name;
                    string ParameterLine;
                    int Index = OptionLine.IndexOf(":");
                    if (Index >= 0)
                    {
                        Name = DescapeQuota(OptionLine.Substring(0, Index));
                        ParameterLine = OptionLine.Substring(Index + 1);
                    }
                    else
                    {
                        Name = DescapeQuota(OptionLine);
                        ParameterLine = "";
                    }
                    Options.Add(new CommandLineOption() { Name = Name, Arguments = (from t in ParameterLine.Split(',') select DescapeQuota(t.Trim(' '))).ToArray() });
                }
                else
                {
                    Arguments.Add(arg);
                }
            }

            return new CommandLineArguments() { Arguments = Arguments.ToArray(), Options = Options.ToArray() };
        }
    }
}